spirv_msl.cpp

```diff
@@ -10682,68 +11957,529 @@ void CompilerMSL::OpCodePreprocessor::check_resource_write(uint32_t var_id)
+			type = &get<SPIRType>(type->member_types[index]);
+		}
+		// Matrix -> Vector
+		else if (type->columns > 1)
+		{
+			// If we have a row-major matrix here, we need to defer any transpose in case this access chain
+			// is used to store a column. We can resolve it right here and now if we access a scalar directly,
+			// by flipping indexing order of the matrix.
+			
+			expr += "[";
+			if (index_is_literal)
+				expr += convert_to_string(index);
+			else
+				expr += to_expression(index, register_expression_read);
+			expr += "]";
+			
+			type_id = type->parent_type;
+			type = &get<SPIRType>(type_id);
+		}
+		// Vector -> Scalar
+		else if (type->vecsize > 1)
+		{
+			string deferred_index;
+			if (row_major_matrix_needs_conversion)
+			{
+				// Flip indexing order.
+				auto column_index = expr.find_last_of('[');
+				if (column_index != string::npos)
+				{
+					deferred_index = expr.substr(column_index);
+					expr.resize(column_index);
+				}
+			}
+			
+			if (index_is_literal && !is_packed && !row_major_matrix_needs_conversion)
+			{
+				expr += ".";
+				expr += index_to_swizzle(index);
+			}
+			else if (ir.ids[index].get_type() == TypeConstant && !is_packed && !row_major_matrix_needs_conversion)
+			{
+				auto &c = get<SPIRConstant>(index);
+				expr += ".";
+				expr += index_to_swizzle(c.scalar());
+			}
+			else if (index_is_literal)
+			{
+				// For packed vectors, we can only access them as an array, not by swizzle.
+				expr += join("[", index, "]");
+			}
+			else
+			{
+				expr += "[";
+				expr += to_expression(index, register_expression_read);
+				expr += "]";
+			}
+			
+			expr += deferred_index;
+			row_major_matrix_needs_conversion = false;
+			
+			is_packed = false;
+			physical_type = 0;
+			type_id = type->parent_type;
+			type = &get<SPIRType>(type_id);
+		}
+		else if (!backend.allow_truncated_access_chain)
+			SPIRV_CROSS_THROW("Cannot subdivide a scalar value!");
+	}
+	
+	if (pending_array_enclose)
+	{
+		SPIRV_CROSS_THROW("Flattening of multidimensional arrays were enabled, "
+						  "but the access chain was terminated in the middle of a multidimensional array. "
+						  "This is not supported.");
+	}
+	
+	if (meta)
+	{
+		meta->need_transpose = row_major_matrix_needs_conversion;
+		meta->storage_is_packed = is_packed;
+		meta->storage_is_invariant = is_invariant;
+		meta->storage_physical_type = physical_type;
+	}
+	
+	return expr;
+}
+
 // Returns an enumeration of a SPIR-V function that needs to be output for certain Op codes.
 CompilerMSL::SPVFuncImpl CompilerMSL::OpCodePreprocessor::get_spv_func_impl(Op opcode, const uint32_t *args)
 {
 	switch (opcode)
 	{
-	case OpFMod:
-		return SPVFuncImplMod;
-
-	case OpFunctionCall:
+	// Storage buffer robustness
+	case OpInBoundsAccessChain:
+	case OpAccessChain:
+	case OpPtrAccessChain:
 	{
-		auto &return_type = compiler.get<SPIRType>(args[0]);
-		if (return_type.array.size() > 1)
+		if (compiler.msl_options.enforce_storge_buffer_bounds)
```

Comment:
FYI. 

For Google's WebGPU work, we too needed a way to force sandboxing of memory accesses. 

We wrote a SPIRV-Tools transform to inject index clamping for all access chains (assuming no variable pointers, just plain logical addressing). The transform does a superset of robust-buffer-access, as it also protects internal memory accesses as well.

In our WebGPU flow on top of Metal, the plan is: parse the SPIR-V, validate it, transform it with this sandboxing pass, then feed it to SPIRV-Cross.

You can play with it via spirv-opt --graphics-robust-access

The code is at https://github.com/KhronosGroup/SPIRV-Tools/blob/master/source/opt/graphics_robust_access_pass.cpp 

I have not compared this implementation with the SPIRV-Tools transform.   But I wanted to make you aware of the other work.   We could probably both benefit from looking at each others test cases.

Reply:
Thanks for the hint. I will have a closer look at this spirv-opt pass next week. Would be definitely great to find a common denominator for as many cases as possible instead of divergence.

Reply:
I'd prefer if robustness checks can be done in SPIRV-Tools. No need to duplicate this effort.